pkcs11sec/pkcs11_security.go

```diff
@@ -0,0 +1,642 @@
+	crypto.SHA1:      {0x30, 0x21, 0x30, 0x09, 0x06, 0x05, 0x2b, 0x0e, 0x03, 0x02, 0x1a, 0x05, 0x00, 0x04, 0x14},
+	crypto.SHA224:    {0x30, 0x2d, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x04, 0x05, 0x00, 0x04, 0x1c},
+	crypto.SHA256:    {0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20},
+	crypto.SHA384:    {0x30, 0x41, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x02, 0x05, 0x00, 0x04, 0x30},
+	crypto.SHA512:    {0x30, 0x51, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x03, 0x05, 0x00, 0x04, 0x40},
+	crypto.MD5SHA1:   {},
+	crypto.RIPEMD160: {0x30, 0x20, 0x30, 0x08, 0x06, 0x06, 0x28, 0xcf, 0x06, 0x03, 0x00, 0x31, 0x04, 0x14},
+}
+
+type Pkcs11Security struct {
+	conf *Config
+	log  *logrus.Entry
+
+	fsec *filesec.FileSecurity
+
+	cert    *tls.Certificate
+	pKey    *PrivateKey
+	pin     *string
+	session p11.Session
+}
+
+type PrivateKey struct {
+	PublicKey  crypto.PublicKey
+	PrivateKey *p11.PrivateKey
+}
+
+func (k *PrivateKey) Public() crypto.PublicKey {
+	return k.PublicKey
+}
+
+// need to handle as many hash types as possible, since this is being used by http/tls driver
+func (k *PrivateKey) Sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	prefix, ok := hashPrefixes[opts.HashFunc()]
+	if !ok {
+		return nil, errors.New("unknown hash function")
+	}
+	mechanism := pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS, nil)
+	input := append(prefix, msg...)
+
+	output, err := k.PrivateKey.Sign(*mechanism, input)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+type Config struct {
+	// CAFile is the file where the trusted CA cert resides
+	CAFile string
+
+	// CertCacheDir is the dir where cached certs reside
+	CertCacheDir string
+
+	// PrivilegedUsers is a list of regular expressions that identity privileged users
+	PrivilegedUsers []string
+
+	// AllowList is a list of regular expressions that identity valid users to allow in
+	AllowList []string
+
+	// DisableTLSVerify disables TLS verify in HTTP clients etc
+	DisableTLSVerify bool
+
+	// AlwaysOverwriteCache supports always overwriting the local filesystem cache
+	AlwaysOverwriteCache bool
+
+	// PKCS11DriverFile points to the dynamic library file to use (usually a .so file)
+	PKCS11DriverFile string
+
+	// PKCS11Slot specifies which slot of the pkcs11 device to use
+	PKCS11Slot uint
+}
+
+func New(opts ...Option) (*Pkcs11Security, error) {
+	p := &Pkcs11Security{}
+
+	for _, opt := range opts {
+		err := opt(p)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if p.conf == nil {
+		return nil, errors.New("configuration not given")
+	}
+
+	if p.log == nil {
+		return nil, errors.New("logger not given")
+	}
+
+	if p.conf.PKCS11DriverFile == "" {
+		return nil, errors.New("pkcs11: PKCS11DriverFile option is required")
+	}
+
+	if p.pin != nil {
+		if err := p.loginToToken(); err != nil {
```

Comment:
mind splitting this over 2 lines? got the same style everywhere else more or less? I know its a go thing to put these on one line but it annoys me no end :)

There's a few of these in the file